File: BackEnd/KawkiWebBusiness/src/main/java/pe/edu/pucp/kawkiweb/bo/ProductosBO.java
package pe.edu.pucp.kawkiweb.bo;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import pe.edu.pucp.kawkiweb.daoImp.ProductosDAOImpl;
import pe.edu.pucp.kawkiweb.model.ProductosDTO;
import pe.edu.pucp.kawkiweb.model.utilProducto.CategoriasDTO;
import pe.edu.pucp.kawkiweb.model.utilProducto.EstilosDTO;
import pe.edu.pucp.kawkiweb.dao.ProductosDAO;
import pe.edu.pucp.kawkiweb.model.UsuariosDTO;

public class ProductosBO {

    private ProductosDAO productoDAO;

    public ProductosBO() {
        this.productoDAO = new ProductosDAOImpl();
    }

    public Integer insertar(String descripcion, CategoriasDTO categoria,
            EstilosDTO estilo, Double precio_venta, UsuariosDTO usuario) {

        try {
            // Validaciones
            if (!validarDatosProducto(descripcion, categoria, estilo, precio_venta,
                    usuario)) {
                System.err.println("Error: Datos de producto inválidos");
                return null;
            }

            ProductosDTO productoDTO = new ProductosDTO();
            productoDTO.setDescripcion(descripcion);
            productoDTO.setCategoria(categoria);
            productoDTO.setEstilo(estilo);
            productoDTO.setPrecio_venta(precio_venta);
            productoDTO.setFecha_hora_creacion(LocalDateTime.now());
            productoDTO.setUsuario(usuario);

            return this.productoDAO.insertar(productoDTO);

        } catch (Exception e) {
            System.err.println("Error al insertar producto: " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }

    public ProductosDTO obtenerPorId(Integer producto_id) {
        try {
            if (producto_id == null || producto_id <= 0) {
                System.err.println("Error: ID de producto inválido");
                return null;
            }
            return this.productoDAO.obtenerPorId(producto_id);

        } catch (Exception e) {
            System.err.println("Error al obtener producto por ID: " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }

    public List<ProductosDTO> listarTodos() {
        try {
            List<ProductosDTO> lista = this.productoDAO.listarTodos();
            return (lista != null) ? lista : new ArrayList<>();

        } catch (Exception e) {
            System.err.println("Error al listar productos: " + e.getMessage());
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    public Integer modificar(Integer producto_id, String descripcion,
            CategoriasDTO categoria, EstilosDTO estilo, Double precio_venta,
            UsuariosDTO usuario) {

        try {
            // Validar ID
            if (producto_id == null || producto_id <= 0) {
                System.err.println("Error: ID de producto inválido");
                return null;
            }

            // Validar datos
            if (!validarDatosProducto(descripcion, categoria, estilo, precio_venta,
                    usuario)) {
                System.err.println("Error: Datos de producto inválidos");
                return null;
            }

            ProductosDTO productoDTO = new ProductosDTO();
            productoDTO.setProducto_id(producto_id);
            productoDTO.setDescripcion(descripcion);
            productoDTO.setCategoria(categoria);
            productoDTO.setEstilo(estilo);
            productoDTO.setPrecio_venta(precio_venta);
            productoDTO.setFecha_hora_creacion(LocalDateTime.now());
            productoDTO.setUsuario(usuario);

            return this.productoDAO.modificar(productoDTO);

        } catch (Exception e) {
            System.err.println("Error al modificar producto: " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }

//    public Integer eliminar(Integer producto_id) {
//        try {
//            if (producto_id == null || producto_id <= 0) {
//                System.err.println("Error: ID de producto inválido");
//                return null;
//            }
//
//            // Verificar si el producto tiene variantes
//            ProductosDTO producto = this.obtenerPorId(producto_id);
//            if (producto != null && producto.getVariantes() != null
//                    && !producto.getVariantes().isEmpty()) {
//                System.err.println("Error: No se puede eliminar un producto con variantes asociadas");
//                return null;
//            }
//
//            ProductosDTO productoDTO = new ProductosDTO();
//            productoDTO.setProducto_id(producto_id);
//            return this.productoDAO.eliminar(productoDTO);
//
//        } catch (Exception e) {
//            System.err.println("Error al eliminar producto: " + e.getMessage());
//            e.printStackTrace();
//            return null;
//        }
//    }
    /**
     * Valida los datos básicos de un producto
     *
     * @return true si los datos son válidos, false en caso contrario
     */
    private boolean validarDatosProducto(String descripcion, CategoriasDTO categoria,
            EstilosDTO estilo, Double precio_venta, UsuariosDTO usuario) {

        // Validar descripción
        if (descripcion == null || descripcion.trim().isEmpty()) {
            System.err.println("Validación: La descripción no puede estar vacía");
            return false;
        }

        if (descripcion.trim().length() > 100) {
            System.err.println("Validación: La descripción es demasiado larga (máx. 500 caracteres)");
            return false;
        }

        // Validar categoría
        if (categoria == null || categoria.getCategoria_id() == null) {
            System.err.println("Validación: La categoría no puede ser null");
            return false;
        }

        // Validar estilo
        if (estilo == null || estilo.getEstilo_id() == null) {
            System.err.println("Validación: El estilo no puede ser null");
            return false;
        }

        // Validar precio de venta
        if (precio_venta == null || precio_venta <= 0) {
            System.err.println("Validación: El precio de venta debe ser mayor a 0");
            return false;
        }

        // Validar usuario
        if (usuario == null || usuario.getUsuarioId() == null) {
            System.err.println("Validación: El usuario no puede ser null");
            return false;
        }

        return true;
    }

    // ========== MÉTODOS OPTIMIZADOS CON STORED PROCEDURES ==========
    /**
     * Verifica si un producto tiene stock disponible en alguna de sus
     * variantes.
     *
     * @param producto_id ID del producto a verificar
     * @return true si tiene stock disponible, false en caso contrario
     */
    public boolean tieneStockDisponible(Integer producto_id) {
        try {
            if (producto_id == null || producto_id <= 0) {
                System.err.println("Error: ID de producto inválido");
                return false;
            }

            return this.productoDAO.tieneStockDisponible(producto_id);

        } catch (Exception e) {
            System.err.println("Error al verificar stock del producto: " + e.getMessage());
            return false;
        }
    }

    /**
     * Calcula el stock total de un producto sumando todas sus variantes.
     *
     * @param producto_id ID del producto
     * @return Stock total, o 0 si hay error
     */
    public Integer calcularStockTotal(Integer producto_id) {
        try {
            if (producto_id == null || producto_id <= 0) {
                System.err.println("Error: ID de producto inválido");
                return 0;
            }

            return this.productoDAO.calcularStockTotal(producto_id);

        } catch (Exception e) {
            System.err.println("Error al calcular stock total: " + e.getMessage());
            return 0;
        }
    }

    /**
     * Lista productos por categoría.
     *
     * @param categoria_id ID de la categoría
     * @return Lista de productos de la categoría
     */
    public List<ProductosDTO> listarPorCategoria(Integer categoria_id) {
        try {
            if (categoria_id == null || categoria_id <= 0) {
                System.err.println("Error: ID de categoría inválido");
                return new ArrayList<>();
            }

            List<ProductosDTO> lista = this.productoDAO.listarPorCategoria(categoria_id);
            return (lista != null) ? lista : new ArrayList<>();

        } catch (Exception e) {
            System.err.println("Error al listar productos por categoría: " + e.getMessage());
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    /**
     * Lista productos por estilo.
     *
     * @param estilo_id ID del estilo
     * @return Lista de productos del estilo
     */
    public List<ProductosDTO> listarPorEstilo(Integer estilo_id) {
        try {
            if (estilo_id == null || estilo_id <= 0) {
                System.err.println("Error: ID de estilo inválido");
                return new ArrayList<>();
            }

            List<ProductosDTO> lista = this.productoDAO.listarPorEstilo(estilo_id);
            return (lista != null) ? lista : new ArrayList<>();

        } catch (Exception e) {
            System.err.println("Error al listar productos por estilo: " + e.getMessage());
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    /**
     * Lista productos con stock bajo (tienen variantes con alerta de stock
     * activada).
     *
     * @return Lista de productos con stock bajo
     */
    public List<ProductosDTO> listarConStockBajo() {
        try {
            List<ProductosDTO> lista = this.productoDAO.listarConStockBajo();
            return (lista != null) ? lista : new ArrayList<>();

        } catch (Exception e) {
            System.err.println("Error al listar productos con stock bajo: " + e.getMessage());
            e.printStackTrace();
            return new ArrayList<>();
        }
    }
}

File: FrontEnd_v2/KawkiWebBusiness/ProductosBO.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using KawkiWebBusiness.KawkiWebWSProductos;
using KawkiWebBusiness.KawkiWebWSUsuarios;
using usuariosDTO = KawkiWebBusiness.KawkiWebWSProductos.usuariosDTO;

namespace KawkiWebBusiness
{
    public class ProductosBO
    {
        private ProductosClient clienteSOAP;

        public ProductosBO()
        {
            this.clienteSOAP = new ProductosClient();
        }

        /// Inserta un nuevo producto en la base de datos
        public int InsertarProducto(string descripcion, categoriasDTO categoria, estilosDTO estilo, 
            double precio_Venta, usuariosDTO usuario)
        {
            //return this.clienteSOAP.insertarProducto(descripcion, categoria, estilo, precioVenta, usuario);
            try
            {
                int resultado = this.clienteSOAP.insertarProducto(descripcion, categoria, estilo, precio_Venta, usuario);

                System.Diagnostics.Debug.WriteLine($"✅ Respuesta del servidor: {resultado}");

                return resultado;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"❌ EXCEPCIÓN en InsertarProducto: {ex.Message}");
                System.Diagnostics.Debug.WriteLine($"StackTrace: {ex.StackTrace}");

                if (ex.InnerException != null)
                {
                    System.Diagnostics.Debug.WriteLine($"Inner Exception: {ex.InnerException.Message}");
                }

                return 0;
            }
        }

        /// Obtiene un producto por su ID
        public productosDTO ObtenerPorIdProducto(int productoId)
        {
            return this.clienteSOAP.obtenerPorIdProducto(productoId);
        }

        /// Lista todos los productos (con sus variantes)
        public IList<productosDTO> ListarTodosProducto()
        {
            return this.clienteSOAP.listarTodosProducto();
        }

        /// Modifica un producto existente
        public int ModificarProducto(int productoId, string descripcion, categoriasDTO categoria, estilosDTO estilo, double precioVenta, usuariosDTO usuario)
        {
            return this.clienteSOAP.modificarProducto(productoId, descripcion, categoria, estilo, precioVenta, usuario);
        }

        /// Elimina un producto por su ID
        //public int? Eliminar(int productoId)
        //{
        //    try
        //    {
        //        if (productoId <= 0)
        //        {
        //            System.Diagnostics.Debug.WriteLine("Error: ID de producto inválido");
        //            return null;
        //        }

        //        // Verificar si el producto tiene variantes
        //        var producto = this.ObtenerPorId(productoId);
        //        if (producto != null && producto.variantes != null && producto.variantes.Length > 0)
        //        {
        //            System.Diagnostics.Debug.WriteLine("Error: No se puede eliminar un producto con variantes asociadas");
        //            return null;
        //        }

        //        int resultado = this.clienteSOAP.eliminarProducto(productoId);
        //        return resultado > 0 ? (int?)resultado : null;
        //    }
        //    catch (Exception ex)
        //    {
        //        System.Diagnostics.Debug.WriteLine($"Error al eliminar producto: {ex.Message}");
        //        return null;
        //    }
        //}

        /// Verifica si un producto tiene stock disponible en alguna de sus variantes
        public bool TieneStockDisponibleProducto(int productoId)
        {
            return this.clienteSOAP.tieneStockDisponibleProducto(productoId);
        }

        /// Calcula el stock total de un producto sumando todas sus variantes
        public int CalcularStockTotalProducto(int productoId)
        {
            return this.clienteSOAP.calcularStockTotalProducto(productoId);
        }

        /// Lista productos por categoría
        public IList<productosDTO> ListarPorCategoriaProducto(int categoriaId)
        {
            try
            {
                if (categoriaId <= 0)
                {
                    return new List<productosDTO>();
                }

                var lista = this.clienteSOAP.listarPorCategoriaProducto(categoriaId);
                if (lista == null) return new List<productosDTO>();
                return new List<productosDTO>(lista);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al listar productos por categoría: {ex.Message}");
                return new List<productosDTO>();
            }
        }

        /// Lista productos por estilo
        public IList<productosDTO> ListarPorEstiloProducto(int estiloId)
        {
            try
            {
                if (estiloId <= 0)
                {
                    return new List<productosDTO>();
                }

                var lista = this.clienteSOAP.listarPorEstiloProducto(estiloId);
                if (lista == null) return new List<productosDTO>();
                return new List<productosDTO>(lista);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al listar productos por estilo: {ex.Message}");
                return new List<productosDTO>();
            }
        }

        /// Lista productos con stock bajo (tienen variantes con alerta de stock activada)
        public IList<productosDTO> ListarConStockBajoProducto()
        {
            try
            {
                var lista = this.clienteSOAP.listarConStockBajoProducto();
                if (lista == null) return new List<productosDTO>();
                return new List<productosDTO>(lista);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al listar productos con stock bajo: {ex.Message}");
                return new List<productosDTO>();
            }
        }
        
    }
}
